File: apps/web/src/lib/discovery.ts
/**
 * اكتشاف المراجع | Reference discovery client.
 *
 * **ثلاثة أشياء لا يجوز أن تصير شيئًا واحدًا**: نتيجةُ بحثٍ في فهرسٍ
 * خارجي، ومرجعٌ مخزَّن في مكتبة الباحث، ودليلٌ يُبنى عليه ادعاء. ولذلك
 * `Candidate` هنا **بلا معرّف في قاعدتنا وبلا حال استعمال**: النوع نفسه
 * يمنع أن يُمرَّر مرشَّحٌ حيث يُنتظر مرجع.
 *
 * والحفظ فعلٌ مستقل يمرّ بالمسار القائم `POST /sources/import` — لا يُعاد
 * بناؤه هنا: مسارٌ ثانٍ للاستيراد يعني قاعدتَي تحقّقٍ تفترقان يومًا.
 */
import { apiFetch } from "./api";
import type { Locale } from "./i18n";

/** ما قاله فهرسٌ واحد — منسوبًا إليه بالاسم. */
export interface ProviderClaim {
  provider: string;
  provider_id: string;
  doi: string | null;
  title: string;
  authors: string[];
  year: number | null;
  venue: string | null;
  volume: string | null;
  issue: string | null;
  pages: string | null;
  abstract: string | null;
  url: string | null;
  open_access: boolean | null;
  citation_count: number | null;
  /** `type` نصّ الفهرس كما قاله، و`work_type` سلّة العرض الموحّدة بين الفهارس. */
  type: string | null;
  work_type: string | null;
  retraction_status: string;
}

/** لماذا اجتمعت ادعاءات الفهارس في بطاقةٍ واحدة — يُعرض ليُراجَع. */
export type MatchBasis = "doi" | "provider_id" | "title_year_author" | "single";

/**
 * سببُ موضع النتيجة في الترتيب، بلغة الباحث.
 *
 * **ولا نسبة في هذا النوع.** «٩٧٪ صلة» رقمٌ لا وحدة له ولا مرجع، ويقرؤه
 * الباحث حكمًا كميًّا على ورقةٍ لم يقرأها. والخادم لا يرسل درجة أصلًا،
 * فلا تستطيع هذه الشاشة اختراع واحدة ولو أرادت.
 */
export interface RankReason {
  code: string;
  /** `match` سببُ ترجيح، و`caution` تنبيهٌ يُقرأ قبل البناء عليه. */
  kind: "match" | "caution" | string;
  terms: string[];
  /** الاستشهاد لا يُذكر بلا قائله: «١٣٤ في OpenAlex». */
  provider: string | null;
  count: number | null;
  year: number | null;
}

/** مصطلحٌ مقترح — معروضٌ لا مُطبَّق. `applied` تعود `false` دائمًا. */
export interface SuggestedTerm {
  term: string;
  source_term: string;
  kind: string;
  applied: boolean;
}

/**
 * ما فهمه الخادم من نصّ الباحث.
 *
 * `raw` سؤاله و`sent` ما غادر إلى الفهارس. يُعرضان معًا حين يختلفان —
 * وهذا هو الحدّ بين «فهمِ الاستعلام» و«إعادةِ كتابته»: ما دام النصّان
 * مرئيَّين، لا يقع تبديلٌ لا يراه صاحبه.
 */
export interface QueryUnderstanding {
  raw: string;
  sent: string;
  doi: string | null;
  phrase: string | null;
  authors: string[];
  year: number | null;
  year_from: number | null;
  year_to: number | null;
  keywords: string[];
  accepted_terms: string[];
  suggestions: SuggestedTerm[];
}

export interface ReferenceCandidate {
  doi: string | null;
  title: string;
  authors: string[];
  year: number | null;
  venue: string | null;
  volume: string | null;
  issue: string | null;
  pages: string | null;
  abstract: string | null;
  url: string | null;
  open_access: boolean | null;
  type: string | null;
  work_type: string | null;
  retraction_status: string;
  providers: string[];
  /** عدّاد كل فهرسٍ منسوبًا إليه — **ولا مجموع**. */
  citation_counts: Record<string, number>;
  match_basis: MatchBasis;
  claims: ProviderClaim[];
  can_be_saved: boolean;
  /** أسبابُ موضعه في الترتيب — **ولا درجة**: الخادم لا يرسل واحدة. */
  reasons: RankReason[];
  matched_terms: string[];
  missing_terms: string[];
}

export interface ProviderStatus {
  provider: string;
  ok: boolean;
  detail: string | null;
  results: number;
}

export interface ExternalAccessLink {
  url: string;
  host: string;
  verified: boolean;
  note_ar: string;
  note_en: string;
}

export interface ReferenceSearchResponse {
  candidates: ReferenceCandidate[];
  providers: ProviderStatus[];
  providers_enabled: boolean;
  any_provider_failed: boolean;
  all_providers_failed: boolean;
  external_link: ExternalAccessLink | null;
  /** بأيّ شيء رُتِّبت النتائج. يُقال صراحةً لأن الباحث يفترض الترتيب الزمني. */
  ordered_by: string;
  query_understanding: QueryUnderstanding | null;
  note_ar: string;
  note_en: string;
}

export interface ReferenceSearchFilters {
  yearFrom?: number | null;
  yearTo?: number | null;
  workType?: string | null;
  openAccessOnly?: boolean;
  /**
   * ما قبِله الباحث من المصطلحات المقترحة — **ولا شيء غيره يوسّع البحث**.
   * وإرسال قائمة فارغة هو الحال الطبيعية: الاقتراح معروضٌ لا مُطبَّق.
   */
  acceptedTerms?: string[];
}

export const searchReferences = (
  locale: Locale,
  query: string,
  filters: ReferenceSearchFilters = {},
) =>
  apiFetch<ReferenceSearchResponse>("/api/v1/references/search", {
    locale,
    method: "POST",
    body: JSON.stringify({
      query,
      limit: 20,
      year_from: filters.yearFrom ?? null,
      year_to: filters.yearTo ?? null,
      work_type: filters.workType || null,
      open_access_only: Boolean(filters.openAccessOnly),
      accepted_terms: filters.acceptedTerms ?? [],
    }),
  });

/** المرجع المخزَّن كما يعيده الخادم — وله معرّفٌ في قاعدتنا، بخلاف المرشَّح. */
export interface StoredSource {
  id: string;
  doi: string | null;
  title: string;
  publication_year: number | null;
  journal_name: string | null;
  verification_status: string;
}

/**
 * احفظ في مكتبتي — **بمعرّفٍ شرعي وحده**.
 *
 * الخادم يحلّ الـDOI في فهرسٍ حقيقي أو يردّ بخطأ؛ ولا يُخزَّن مرجعٌ
 * مختلق. ومرشَّحٌ بلا DOI لا يُحفظ أصلًا — وذلك يُقال للباحث في الشاشة،
 * ولا يُسدّ بمعرّفٍ يُصنع له.
 */
export const saveToLibrary = (locale: Locale, doi: string) =>
  apiFetch<StoredSource>("/api/v1/sources/import", {
    locale,
    method: "POST",
    body: JSON.stringify({ doi }),
  });

File: apps/api/athera_api/schemas/discovery.py
"""عقود اكتشاف المراجع | Reference discovery contracts.

**المرشَّح ليس مرجعًا مخزَّنًا، والمرجع المخزَّن ليس دليلًا.** ثلاث حالات
لا يجوز طيّها في واحدة، والعقد هنا يحفظ الأولى منها: لا معرّف لمرشَّح في
قاعدتنا، ولا حال استعمال — لأنه لم يدخل المكتبة بعد.

وكل حقلٍ يبقى مصحوبًا بمن قاله في `claims`؛ فالقيمة المعروضة اختيارٌ
بأسبقيةٍ مُعلَنة، لا حقيقةٌ بلا مصدر.
"""
from __future__ import annotations

from pydantic import BaseModel, Field


class ReferenceSearchRequest(BaseModel):
    """عنوانٌ أو كلماتٌ مفتاحية أو DOI — والخادم يميّزها، لا الباحث."""

    query: str = Field(min_length=2, max_length=1000)
    limit: int = Field(default=20, ge=1, le=50)
    year_from: int | None = Field(default=None, ge=1400, le=2200)
    year_to: int | None = Field(default=None, ge=1400, le=2200)
    work_type: str | None = Field(default=None, max_length=64)
    open_access_only: bool = False


class ProviderClaimView(BaseModel):
    """ما قاله فهرسٌ واحد — منسوبًا إليه بالاسم."""

    provider: str
    provider_id: str
    doi: str | None = None
    title: str
    authors: list[str] = []
    year: int | None = None
    venue: str | None = None
    volume: str | None = None
    issue: str | None = None
    pages: str | None = None
    abstract: str | None = None
    url: str | None = None
    open_access: bool | None = None
    citation_count: int | None = None
    type: str | None = None
    retraction_status: str


class ReferenceCandidateView(BaseModel):
    """مرشَّح بحثٍ واحد. الحقول للعرض، و`claims` للنسبة."""

    doi: str | None = None
    title: str
    authors: list[str] = []
    year: int | None = None
    venue: str | None = None
    volume: str | None = None
    issue: str | None = None
    pages: str | None = None
    abstract: str | None = None
    url: str | None = None
    open_access: bool | None = None
    type: str | None = None
    retraction_status: str
    providers: list[str] = []
    # عدّاد كل فهرسٍ منسوبًا إليه — لا مجموع ولا متوسّط.
    citation_counts: dict[str, int] = {}
    # لماذا اجتمعت هذه الادعاءات في بطاقةٍ واحدة: doi | provider_id |
    # title_year_author | single. يجعل الدمج مُراجَعًا لا مسلَّمًا به.
    match_basis: str
    claims: list[ProviderClaimView] = []
    # الحفظ في المكتبة يمرّ بمعرّفٍ شرعيّ. وبلا DOI لا يُخزَّن مرجعٌ متحقَّق
    # ولا يُختلق له معرّف — فيُقال ذلك في العقد بدل أن يفشل الزرّ صامتًا.
    can_be_saved: bool = False


class ProviderStatusView(BaseModel):
    """حال فهرسٍ في هذه التشغيلة — يُعرض ليُقرأ الفشل فشلًا."""

    provider: str
    ok: bool
    detail: str | None = None
    results: int = 0


class ExternalAccessLinkView(BaseModel):
    """رابط وصول إضافي — يُحفظ ولا يُقرأ منه شيء.

    `verified` هنا `False` دائمًا وبنيةً: المنصّة لم تُسأل ولن تُسأل.
    """

    url: str
    host: str
    verified: bool = False
    note_ar: str = (
        "هذا رابط وصول إضافي فقط. لا تُجمع منه بيانات وصفية ولا يُعدّ متحقَّقًا؛ "
        "البيانات تأتي من معرّفٍ شرعي أو من فهرسٍ علمي."
    )
    note_en: str = (
        "This is an additional access link only. No metadata is collected from it and it "
        "is not treated as verified; metadata comes from a legitimate identifier or index."
    )


class ReferenceSearchResponse(BaseModel):
    """نتيجة الاكتشاف — ومعها من أجاب ومن تعذّر، لأن الفرق بينهما جوهري."""

    candidates: list[ReferenceCandidateView] = []
    providers: list[ProviderStatusView] = []
    # `False` تعني أن هذا النشر لم يُفعّل فيه فهرسٌ خارجي أصلًا — وهي حالٌ
    # ثالثة لا تُقرأ «لا نتائج» ولا «تعذّر الفهرس».
    providers_enabled: bool = True
    any_provider_failed: bool = False
    all_providers_failed: bool = False
    external_link: ExternalAccessLinkView | None = None
    note_ar: str = "نتيجة البحث ليست مرجعًا مخزَّنًا، والمرجع المخزَّن ليس دليلًا."
    note_en: str = (
        "A search result is not a stored reference, and a stored reference is not evidence."
    )
